cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,869 @@
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlUserTaskResponse;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlUserTaskStatus;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.RebalanceOptions;
+import io.strimzi.operator.cluster.operator.resource.ResourceOperatorSupplier;
+import io.strimzi.operator.common.AbstractOperator;
+import io.strimzi.operator.common.Annotations;
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.Util;
+import io.strimzi.operator.common.model.Labels;
+import io.strimzi.operator.common.operator.resource.AbstractWatchableResourceOperator;
+import io.strimzi.operator.common.operator.resource.CrdOperator;
+import io.strimzi.operator.common.operator.resource.StatusUtils;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+import static io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlApi.CC_REST_API_SUMMARY;
+
+/**
+ * <p>Assembly operator for a "KafkaRebalance" assembly, which interacts with Cruise Control REST API</p>
+ *
+ * <p>
+ *     This operator takes care of the {@code KafkaRebalance} custom resources that a user can create in order
+ *     to interact with Cruise Control REST API and execute a cluster rebalancing.
+ *     A state machine is used for the rebalacing flow which is reflected in the {@code status} of the custom resource.
+ *
+ *     When a new {@code KafkaRebalance} custom resource is created, the operator sends a rebalance proposal
+ *     request to the Cruise Control REST API and moves to the {@code PendingProposal} state. It stays in this state
+ *     until a the rebalance proposal is ready, polling the related status on Cruise Control, and then finally moves
+ *     to the {@code ProposalReady} state. The status of the {@code KafkaRebalance} custom resource is updated with the
+ *     computed rebalance proposal so that the user can view it and making a decision to execute it or not.
+ *     For starting the actual rebalancing on the cluster, the user annotate the custom resource with
+ *     the {@code strimzi.io/rebalance=approve} annotation, triggering the operator to send a rebalance request to the
+ *     Cruise Control REST API in order to execute the rebalancing.
+ *     During the rebalancing, the operator state machine is in the {@code Rebalancing} state and it moves finally
+ *     to the {@code Ready} state when the rebalancing is done.
+ *
+ *     The user is also able to stop an in-progress rebalance proposal computation or an actual rebalancing,
+ *     annotating the custom resource with {@code strimzi.io/rebalance=stop} when it is in {@code PendingProposal}
```

Comment:
So stopping a in-progress rebalance proposal is not really a thing. There is no way on the CC REST API stopping something like that, stopping here is useful to stop the periodic time for requesting status to CC REST API. About rebalancing, yes there is a way for stopping an ongoing rebalancing task and I guess that it leaves the work as it is. Is that right @tomncooper ?

Reply:
> So stopping a in-progress rebalance proposal is not really a thing.

This Javadoc says it does - it says you can use this annotation when the state is `Rebalancing`.

Reply:
So you can stop an on going `rebalance(dryryun=false)`, you cannot stop the creation of a optimization proposal (`rabalance(dryrun=true)`). As Paolo said the stop annotation here is to signal to the operator that you don't want it to hang around for the proposal to be ready.

Reply:
I guess now I'm totally confused ... you say _cannot stop the creation of a optimization proposal_ but also that it is _to signal to the operator that you don't want it to hang around for the proposal to be ready_  ... isn't creation of the proposal and waiting for the proposal to be ready the same thing?

Reply:
So a user deploys a `rebalance.yaml` with a bunch of custom options. They apply it and off CC goes calculating the optimization while we hold the lock. The user then realises "dam I missed something" and stops the rebalance. If we did not allow for stopping a proposal (on the Strimzi side --- it is not possible on the CC side) the CO would wait for the custom proposal to be ready and then wait for the next reconcile before stopping things and releasing the lock.

Reply:
Ok, that sounds reasonable. So perhaps you need to clarify here (+ somewhere on the docs, but that migth be other PR) what does do something in CC and what does just do something in the operator since that does not seem to be clear. Even now I still don't understand whether stopping an active rebalance is now again operator only or if that is actually in CC.

Reply:
stopping an active rebalance is an actual operation on CC

Reply:
Ok ... since you are using single annotation for two different things, could you make it clear in the comment? (and in the docs?)

Reply:
I updated the comments with clarification of the two stop conditions.